File: 27_09_2024_590017213_Prem_Shandilya.java
class TreeNode {
    int val;
    TreeNode left, right;

    TreeNode(int x) {
        val = x;
        left = right = null;
    }
}

class DoublyLinkedListNode {
    int val;
    DoublyLinkedListNode prev, next;

    DoublyLinkedListNode(int x) {
        val = x;
        prev = next = null;
    }
}

public class BSTToSortedCircularDLL {
    private DoublyLinkedListNode head = null;
    private DoublyLinkedListNode tail = null;

    public DoublyLinkedListNode convert(TreeNode root) {
        if (root == null) return null;

        inOrderTraversal(root);

        if (head != null && tail != null) {
            head.prev = tail;
            tail.next = head;
        }

        return head;
    }

    private void inOrderTraversal(TreeNode node) {
        if (node == null) return;

        inOrderTraversal(node.left);

        DoublyLinkedListNode newNode = new DoublyLinkedListNode(node.val);

        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            newNode.prev = tail;
            tail = newNode;
        }

        inOrderTraversal(node.right);
    }

    public void printList(DoublyLinkedListNode head) {
        if (head == null) return;

        DoublyLinkedListNode current = head;
        do {
            System.out.print(current.val + " ");
            current = current.next;
        } while (current != head);
        System.out.println();
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(2);
        root.right = new TreeNode(6);
        root.left.left = new TreeNode(1);
        root.left.right = new TreeNode(3);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(7);

        BSTToSortedCircularDLL converter = new BSTToSortedCircularDLL();
        DoublyLinkedListNode head = converter.convert(root);
        converter.printList(head);
    }
}

File: 27_09_2024_590017213_Prem_Shandilya.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def binary_tree_to_doubly(root):
    def convert_to_dll(node):
        nonlocal prev, head
        if not node:
            return
        convert_to_dll(node.left)

        if prev is None:
            head = node
        else:
            node.left = prev
            prev.right = node
        
        prev = node
        
        convert_to_dll(node.right)

    prev = None
    head = None
    
    convert_to_dll(root)
    
    return head

def print_dll(head):
    while head:
        print(head.data, end=" ")
        head = head.right
    print()

if __name__ == "__main__":
    root = Node(13)
    root.left = Node(7)
    root.right = Node(24)
    root.left.left = Node(6)
    root.left.right = Node(7)
    root.right.left = Node(18)
    root.right.right = Node(28)
    
    dll_head = binary_tree_to_doubly(root)
    print("Doubly Linked List in sorted order:")
    print_dll(dll_head)
